src/nbl/ext/TextRendering/TextRendering.cpp

```diff
@@ -0,0 +1,250 @@
+
+using namespace nbl;
+using namespace nbl::core;
+using namespace nbl::asset;
+using namespace nbl::video;
+
+#include "nabla.h"
+#include <nbl/ext/TextRendering/TextRendering.h>
+
+namespace nbl
+{
+namespace ext
+{
+namespace TextRendering
+{
+
+uint32_t getCharacterAtlasPositionIx(char character)
+{
+	return int(character) - int(' ');
+}
+
+// Generates atlas of MSDF textures for each ASCII character
+FontAtlas::FontAtlas(IGPUQueue* queue, ILogicalDevice* device, const std::string& fontFilename, uint32_t msdfWidth, uint32_t msdfHeight, uint32_t charsPerRow)
+{
+	auto error = FT_Init_FreeType(&library);
+	assert(!error);
+
+	error = FT_New_Face(library, "C:\\Windows\\Fonts\\arialbd.ttf", 0, &face);
+	assert(!error);
+
+	auto fence = device->createFence(static_cast<nbl::video::IGPUFence::E_CREATE_FLAGS>(0));
+	auto commandPool = device->createCommandPool(queue->getFamilyIndex(), nbl::video::IGPUCommandPool::ECF_NONE);
+	nbl::core::smart_refctd_ptr<nbl::video::IGPUCommandBuffer> commandBuffer;
+	device->createCommandBuffers(commandPool.get(), nbl::video::IGPUCommandBuffer::EL_PRIMARY, 1u, &commandBuffer);
+
+	std::vector<nbl::core::smart_refctd_ptr<video::IGPUBuffer>> dataBuffers;
+	std::vector<asset::IImage::SBufferCopy> bufferCopies;
+
+	uint32_t characterCount = 0;
+	{
+		uint32_t atlasGlyphX = 0;
+		uint32_t atlasRow = 0;
+
+		for (char k = ' '; k <= '~'; k++)
+		{
+			wchar_t unicode = wchar_t(k);
+			uint32_t glyphIndex = FT_Get_Char_Index(face, unicode);
+
+			// special case for space as it seems to break msdfgen
+			if (glyphIndex == 0 || k == ' ')
+				continue;
+
+			error = FT_Load_Glyph(face, glyphIndex, FT_LOAD_NO_SCALE);
+			assert(!error);
+
+			msdfgen::Shape shape;
+			bool loadedGlyph = nbl::ext::TextRendering::TextRenderer::getGlyphShape(shape, library, face);
+			assert(loadedGlyph);
+
+			shape.normalize();
+			msdfgen::edgeColoringSimple(shape, 3.0); // TODO figure out what this is
+			msdfgen::Bitmap<float, 3> msdfMap(msdfWidth, msdfHeight);
+			msdfgen::generateMSDF(msdfMap, shape, 4.0, 1.0, { 4.0, 4.0 });
+
+			uint32_t rowLength = msdfWidth * 4;
+
+			video::IGPUBuffer::SCreationParams bufParams;
+			bufParams.size = rowLength * msdfHeight;
+			bufParams.usage = asset::IBuffer::EUF_TRANSFER_SRC_BIT;
+
+			auto data = device->createBuffer(std::move(bufParams));
```

Comment:
TODO: Merge with `image_upload_utils` and use `uploadImageViaStagingBuffer` no need for this buffer anymore, use most explicit the variation where you create the fence, cmdbuffer and submit it yourself so you can record all the atlas copies and then submit at once.

(you would need to access an IUtilities object)

Reply:
there is a single submit in the current implementation, im not too familiar with image_upload_utils to know what you mean

Reply:
That's ok, just put the TODO.

See how I use it  [here](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/blob/57a9f76a17d8659f5df471487c35771a27b098c3/09.ColorSpaceTest/main.cpp#L421)


that's the tool that helps you upload your CPU Buffer to a GPUImage region in chunks. It can also convert the format while copying.